skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc

```diff
@@ -0,0 +1,424 @@
+            D3D_FEATURE_LEVEL_12_0
+        };
+
+        for (int i = 0; i < _countof(featureLevels); i++) {
+            if (SUCCEEDED(D3D12CreateDevice(adapter.get(), featureLevels[i], _uuidof(ID3D12Device), nullptr))) {
+                maxSupportedFeatureLevel = featureLevels[i];
+                break;
+            }
+        }
+
+        gr_cp<ID3D12Device> device;
+        if (!SUCCEEDED(D3D12CreateDevice(adapter.get(), maxSupportedFeatureLevel, IID_PPV_ARGS(&device)))) {
+            return 0;
+        }
+
+        // Create the command queue
+        gr_cp<ID3D12CommandQueue> queue;
+        D3D12_COMMAND_QUEUE_DESC queueDesc = {};
+        queueDesc.Flags = D3D12_COMMAND_QUEUE_FLAG_NONE;
+        queueDesc.Type = D3D12_COMMAND_LIST_TYPE_DIRECT;
+
+        if (!SUCCEEDED(device->CreateCommandQueue(&queueDesc, IID_PPV_ARGS(&queue)))) {
+            return 0;
+        }
+
+        ID3D12CommandAllocator* commandAllocator;
+        if (!SUCCEEDED(device->CreateCommandAllocator(D3D12_COMMAND_LIST_TYPE_DIRECT, IID_PPV_ARGS(&commandAllocator)))) {
+            return 0;
+        }
+
+        ID3D12GraphicsCommandList* commandList;
+        if (!SUCCEEDED(device->CreateCommandList(0, D3D12_COMMAND_LIST_TYPE_DIRECT, commandAllocator, nullptr, IID_PPV_ARGS(&commandList)))) {
+            return 0;
+        }
+
+        ID3D12Fence* fence;
+        if (!SUCCEEDED(device->CreateFence(0, D3D12_FENCE_FLAG_NONE, IID_PPV_ARGS(&fence)))) {
+            return 0;
+        }
+
+        HANDLE fenceEvent = CreateEventEx(nullptr, false, false, EVENT_ALL_ACCESS);
+        if (!fenceEvent) {
+            return 0;
+        }
+
+        DirectXOffscreenDevice *d3dDevice = new DirectXOffscreenDevice();
+        d3dDevice->commandAllocator = commandAllocator;
+        d3dDevice->commandList = commandList;
+        d3dDevice->fence = fence;
+        d3dDevice->fenceEvent = fenceEvent;
+        d3dDevice->backendContext.fAdapter = adapter;
+        d3dDevice->backendContext.fDevice = device;
+        d3dDevice->backendContext.fQueue = queue;
+        d3dDevice->backendContext.fProtectedContext = GrProtected::kNo;
+
+        return toJavaPointer(d3dDevice);
+    }
+
+    JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_swing_Direct3DSwingRedrawer_makeDirectXRenderTargetOffScreen(
+            JNIEnv *env, jobject redrawer, jlong texturePtr) {
+        DirectXOffScreenTexture *texture = fromJavaPointer<DirectXOffScreenTexture *>(texturePtr);
+        ID3D12Resource* resource = texture->resource;
+
+        GrD3DTextureResourceInfo texResInfo = {};
+        texResInfo.fResource.retain(resource);
+        texResInfo.fResourceState = D3D12_RESOURCE_STATE_COMMON;
+        texResInfo.fFormat = DXGI_FORMAT_R8G8B8A8_UNORM;
+        texResInfo.fSampleCount = 1;
+        texResInfo.fLevelCount = 1;
+        GrBackendRenderTarget* renderTarget = new GrBackendRenderTarget(texture->width, texture->height, texResInfo);
+        return reinterpret_cast<jlong>(renderTarget);
```

Comment:
Both `resource` and `readbackBufferResource` allocations can fail inside DirectXOffScreenTexture constructor, we need to check that both are not nullptr or fallback otherwise (delete DirectXOffScreenTexture and return zero).